src/Tutor.Core/DomainModel/KnowledgeComponents/KnowledgeComponentMastery.cs

```diff
@@ -39,10 +41,43 @@ public KnowledgeComponentMastery(KnowledgeComponent knowledgeComponent)
             KnowledgeComponent = knowledgeComponent;
             IsPassed = false;
             IsSatisfied = false;
+            HasActiveSession = false;
+        }
+
+        public void LaunchSession()
+        {
+            if (HasActiveSession)
+                Causes(new SessionAbandoned()
+                {
+                    KnowledgeComponentId = KnowledgeComponent.Id,
+                    LearnerId = LearnerId
+                });
```

Comment:
I generated this event now to cover cases where termination didn't happen, and it seems to cover that well enough, but note that this won't actually work correctly in cases where an earlier session is actually active.

For example, say a Learner has a KC open in one browser tab, and there is an active session for that. If a Learner opens the same KC in another tab and a new session is launched, a SessionAbandoned event will be generated for the old session, but the Learner won't know that and will actually still be able to interact through the old session. I feel like handling this properly will require adding Ids to sessions or something like that, and connecting all requests to them, so that they can be rejected in cases like this one.